File: controller/src/controller.py
#!/usr/bin/env python

import rospy
from geometry_msgs.msg import PoseStamped
from ackermann_msgs.msg import AckermannDriveStamped

# end point:
#   - connected chain from vrpn->state_estimator->controller->ackermann_mux
#   - controller commands executed by car (manual input topics need to be empty)

#   - need to be able to interpret PoseStamped in controller context
#   - swap hat object for marmot object on vrpn
#   - want to publish commands based on vrpn measurement

# ISSUE: stopping and starting motion
#   - robot stops even when ped is >1, randomly starts again
#   - check if controls are still publishing

class Controller:

    def __init__(self):
        self.ctrl_pub = rospy.Publisher("/car/mux/ackermann_cmd_mux/input/controller", 
            AckermannDriveStamped, 
            queue_size=2)

        self.state_est_sub = rospy.Subscriber("/car/state_estimator/state_estimator/inferred_pose", 
            PoseStamped,
            self.callback)

    def callback(self, pose_msg):
        self.compute_action(pose_msg)

    def compute_action(self, pose_msg):
        x_ped = pose_msg.pose.position.x
        y_ped = pose_msg.pose.position.y
	theta_ped = pose_msg.pose.orientation

	#  print(theta_ped)

        if y_ped > 0.75:
            ctrl = [0.5,0.0]
        elif y_ped < -0.75:
            ctrl = [-0.5,-0.0]
        else:
            ctrl = [0.0,0.0]
            
        self.publish_ctrl(ctrl)

    def publish_ctrl(self, ctrl):
        dur = rospy.Duration(1.0)
        rate = rospy.Rate(10)
        start = rospy.Time.now()

        ctrl_msg = AckermannDriveStamped()
        ctrl_msg.header.stamp = rospy.Time.now()

        ctrl_msg.drive.speed = float(ctrl[0])
        ctrl_msg.drive.steering_angle = float(ctrl[1])

        print(ctrl_msg)
        self.ctrl_pub.publish(ctrl_msg)


if __name__ == '__main__':
    try:
        rospy.init_node("controller")
        c = Controller()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass

File: controller/src/controller.jl
#!/usr/bin/env julia

# RobotOS.jl docs: https://jdlangs.github.io/RobotOS.jl/latest/

using RobotOS
using BSON: @load
using ReferenceFrameRotations

include("/home/adcl/Documents/marmot-algs/HJB-planner/HJB_functions.jl")

@rosimport ackermann_msgs.msg: AckermannDriveStamped
@rosimport geometry_msgs.msg: PoseStamped
@rosimport state_estimator.srv: EstState

rostypegen()
using .ackermann_msgs.msg
using .geometry_msgs.msg
using .state_estimator.srv

# TO-DO: need to propagate state forward one step when choosing new action

function main()
    @load "/home/adcl/Documents/marmot-algs/HJB-planner/bson/U_HJB.bson" U_HJB
    @load "/home/adcl/Documents/marmot-algs/HJB-planner/bson/env.bson" env
    @load "/home/adcl/Documents/marmot-algs/HJB-planner/bson/veh.bson" veh

    init_node("controller")
    ctrl_pub = Publisher{AckermannDriveStamped}(
        "/car/mux/ackermann_cmd_mux/input/controller",
        queue_size=2)

    println("controller node initialized")

    end_run = false
    while end_run == false
        end_run = run_control(ctrl_pub, U_HJB, env, veh) 
    end
end

# NOTE: is this re-initializing the proxy every time? -> possibly, but doesn't seem to matter
function state_estimator_client()
    wait_for_service("/car/state_estimator/get_est_state")
    est_state_srv = ServiceProxy{EstState}("/car/state_estimator/get_est_state")

    req = EstStateRequest(1.0)
    resp = est_state_srv(req)

    s = [resp.x, resp.y, resp.theta]

    return s
end

# TO-DO: need to delegate publish_control to another node, so that this node can be calculating next action during dt
function run_control(ctrl_pub, U_HJB, env, veh)    
    dt = 0.1
    
    s = state_estimator_client()
    println("s: ", s)

    if in_target_set(s, env, veh) == false && in_workspace(s, env, veh) == true
        a = optimal_action_HJB(s, U_HJB, env, veh)

        a[1] = 0.5*a[1]  # TO-DO: change back
        end_run = false
    else
        a = [0.0, 0.0]
        end_run = true
    end

    println("a: ", a)
    println("")

    publish_ctrl(ctrl_pub, a, dt)

    return end_run
end

function publish_ctrl(ctrl_pub, ctrl, dt)
    dur = Duration(dt)
    rate = Rate(10)
    start = get_rostime()

    ctrl_msg = AckermannDriveStamped()
    ctrl_msg.header.stamp = get_rostime()

    ctrl_msg.drive.speed = ctrl[1]
    ctrl_msg.drive.steering_angle = ctrl[2]

    while get_rostime()-start < dur
        publish(ctrl_pub, ctrl_msg)
        sleep(rate)
    end
end

main()